### ISSUE TITLE
Proposal: fork loops

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#8014]
The problem with `foreach` and `for` loops is that they don't play well with `await` and just get blocked until the awaitable returns. I want to propose `fork` loops which don't, similar to fork-join model.

For example,

``` cs
// waits in each iteration
async Task SaveAllAsync(Foo[] data) {
    foreach(var item in data) {
        await item.SaveAsync();
    }
}

// this won't get blocked and will wait in parallel
async Task SaveAllAsync(Foo[] data) {
    fork(var item in data) {
        await item.SaveAsync();
    } // join
}
```

This behaves like a list of running tasks and a `Task.WhenAll` but a little nicer, it provides language support for [processing tasks as they complete](http://blogs.msdn.com/b/pfxteam/archive/2012/08/02/processing-tasks-as-they-complete.aspx) which is a common pattern, I guess.

Note that in each iteration, we don't move to the next item until we get to an `await`, for the sake of fairness we will do even if the task has already been completed.

Iterations won't run in parallel but might be in waiting state at the same time, meaning that we wait for all the tasks in parallel but we'll process them in order — a completed task might be waiting for another iteration to get to an `await` or end of the loop block, this makes sense because all of them will join at the end of the loop.

Similar to `async` which would be redundant without any `await`, `fork` loops have no effect without any `await` in the body, so in that case a warning would be useful.

With #261, one might use `yield return` inside the loop to effectively return a push-based collection,

``` cs
async IObservable<Result> ProcessAllAsync(Foo[] data) {
    fork(var foo in data) {
        yield return await foo.ProcessAsync();
    }
}
```

In contrast, you can use a regular `foreach` to return an `IAsyncEnumerable<T>`,

``` cs
async IAsyncEnumerable<Result> ProcessAllAsync(Foo[] data) {
    foreach(var foo in data) {
        yield return await foo.ProcessAsync();
    }
}
```

Also it should be able to be used to iterate both `IObservable<T>` and `IAsyncEnumerable<T>` collections.

The advantage of this approach compared to `Task.WhenAll` is that only one `Task` at a time is in `Running` state so the body of the loop doesn't "run" in parallel, rather, all the tasks can be "waiting" at the same time.

**Comment by svick** [comment]
This sounds to me like it's embedding a significant amount of policy (execute asynchronous parts with unlimited degree of concurrency, but serialize synchronous parts) into the language, which means it would be useful only in pretty specific cases.

Also, what you're describing sounds to me like `Task.WhenAll()` combined with running on a synchronized `TaskScheduler` (e.g. `ConcurrentExclusiveSchedulerPair.ExclusiveScheduler`) would achieve the same effect.

There certainly are cases where a construct like this one would be useful, and there are also cases that call for different constructs (like parallel asynchronous loop with limited degree of concurrency for both async and sync parts). Considering that, I would like to see improved handling of `async` loops in .Net, but I think that a library (probably similar to `Parallel.ForEach()`) would be a better solution, especially since it could more easily serve the different cases.

**Comment by alrz** [comment]
@svick The point is to seamlessly support async sequences in the language (iterators, loops, etc), If that wasn't the case, this is not necessarily solve any problem. If we ever have async sequence support (which currently Rx provides it with extensive flexibility) I think this kind of constructs can be useful. So you are also against language support for async sequences, right? Because, well, Rx "could more easily serve the different cases".

**Comment by HaloFour** [comment]
@alrz So just that I understand this the only real difference between `fork` and a form of `foreach` that will support `IAsyncEnumerable<T>` is that the former may switch between the completed task contexts within the loop body if it yields on `await`?  So given the following:

```cs
IAsyncEnumerable<int> ids = LookupIdsAsync();
fork (int id in ids) {
    Console.WriteLine($"Looking up user ID {id}.");
    User user = await LookupUserAsync(id);
    Console.WriteLine($"Found user {user.Name} for ID {id}.");
}
Console.WriteLine("Finished looking up users.");
```

The possible output might be the following since the multiple calls to `LookupUserAsync` could be dispatched concurrently:

```
Looking up user ID 1.
Looking up user ID 2.
Found user Bar for ID 2.
Found user Foo for ID 1.
Looking up user ID 3.
Found user Baz for ID 3.
Finished looking up users.
```

Whereas with an async foreach it would be the following since the async method would have to complete the loop body before iterating to the next element in the sequence:

```
Looking up user ID 1.
Found user Foo for ID 1.
Looking up user ID 2.
Found user Bar for ID 2.
Looking up user ID 3.
Found user Baz for ID 3.
Finished looking up users.
```

**Comment by HaloFour** [comment]
I'll also note that there didn't seem to be much support for `IObservable<T>` in #261.

**Comment by alrz** [comment]
@HaloFour First part of the loop runs synchronously until it gets to the `await`, just like any other `async` method. So the output might be like,
```
Looking up user ID 1.
Looking up user ID 2.
Looking up user ID 3.
Found user Bar for ID 2.
Found user Foo for ID 1.
Found user Baz for ID 3.
Finished looking up users.
```
Due to the fairness policy. But after `await` it depends on `LookupUserAsync` for each call, and generally it might complete faster because it kicks all the tasks at first and then wait for each one. The reason behind not allowing any part of the loop to run in parallel is that then you don't need to worry about synchronizing shared state after `await` (note that this doesn't apply to `async` methods in general because of potential reentrancy, but this can be prevented in `fork` loop body).

Yes it's kind of push-based computation and would be more useful if we ever have `IObservable<T>` support.

**Comment by HaloFour** [comment]
@alrz What if the `LookupUserAsync` task completed before the entire asynchronous sequence was finished being iterated?  In that case could it be possible for a "Found user" message to be printed prior to a "Looking up user" message?

**Comment by alrz** [comment]
@HaloFour No. As I said "in each iteration, we don't move to the next item until we get to an `await`, for the sake of fairness we will do even if the task has already been completed."

**Comment by HaloFour** [comment]
@alrz  The call to `LookupUserAsync` would be that `await`.  The call to `MoveNext` on the sequence would be another `await`.  Are you saying that if the task from `LookupUserAsync` completed before the task from `MoveNext` that `fork` would not continue the loop body?

**Comment by alrz** [comment]
@HaloFour Oh I see, that's another story. To quote @tpetricek from [#261 comment](https://github.com/dotnet/roslyn/issues/261#issuecomment-95269316):
>I find `IAsyncEnumerable<T>` quite odd because it lets you call `MoveNext` repeatedly without waiting for the completion of the first - this is probably never desirable

To not get into that situation, when we are waiting for `MoveNextAsync` we don't go further and if any other tasks from previous iterations have been completed we will continue the loop body. So yes, in case of `IAsycEnumerable<T>` it'll be possible for a "Found user" message to be printed prior to a "Looking up user" message.

**Comment by HaloFour** [comment]
@alrz Sure, any form of language-provided asynchronous enumeration should avoid calling `MoveNextAsync` without waiting for the previous invocation's task to complete.

I still have reservations regarding locals, but I do think that a language-provided enumeration like this would be very useful.  This is particularly true if the asynchronous method performing the enumeration is itself an async iterator.  While PFX and `Task.WhenAll` do easily allow synchronization on the individual tasks, they couldn't allow `yield`ing the processed results as they are completed very easily.  LINQ extensions for `IAsyncEnumerable<T>` would make that easier by composing a pipeline, but that's still not as flexible as being able to `yield` arbitrarily within the body of this loop.  And manually writing the enumerating code wouldn't permit resuming based on whichever operation had the next result.

This feels a lot like dispatching a pipeline of operations to fibers.  Not sure I care for `fork`, though, but I don't have a better term to suggest at the moment.

**Comment by alrz** [comment]
@HaloFour "any form of language-provided asynchronous enumeration should avoid calling MoveNextAsync without waiting for the previous invocation's task to complete." they will if they simply use `await`, because other loops don't continue when they are waiting for a task and they'll just get back to the caller. And locals, declared outside of the loop or otherwise, will be safe to use, I can't think of any situation where this could be problematic. PS: Other than compiler generated state machine, I can't think of any other solution to implement this behavior. :smile:

**Comment by svick** [comment]
@alrz  My opinion is that Rx is already good enough for push async sequences, but that pull async sequences should be supported too and that would very likely require language support.

It seems to me that this proposal is about supporting a certain pretty specific pattern of producing `IObservable` directly in the language, which I think could and should be handled by Rx itself.

Actually, you can already do that:

```c#
IObservable<Result> ProcessAllAsync(Foo[] data)
{
    return data.ToObservable()
        .ObserveOn(new TaskPoolScheduler(new TaskFactory(
            new ConcurrentExclusiveSchedulerPair().ExclusiveScheduler)))
        .SelectMany(foo => foo.ProcessAsync());
}
```

That `ObserveOn()` line is horrible, but it should be simple to improve Rx to better support this.

**Comment by alrz** [comment]
@svick
>```cs
IObservable<Result> ProcessAllAsync(Foo[] data)
{
    return data.ToObservable()
        .ObserveOn(new TaskPoolScheduler(new TaskFactory(
            new ConcurrentExclusiveSchedulerPair().ExclusiveScheduler)))
        .SelectMany(foo => foo.ProcessAsync());
}
```

wow. Also, @HaloFour's example was not returning an `IObservable<T>`. By the way, `fork` doesn't mind if you are working with a `IObservable<T>`, `IAsyncEnumerable<T>` or `T[]` or even `yield`. The fact that "it just works"  doesn't make it versatile enough?

**Comment by HaloFour** [comment]
@alrz 

> they will if they simply use `await`

That it would, but to bring a real benefit it would have to invoke the next `MoveNextAsync` effectively immediately after the previous task completed and the current value extracted and then keep track of a list of outstanding operations resuming the coroutine for whichever task completed next.  That way multiple async operations would be in flight and the "fiber" that is coordinating the loop would be jumping to the statement immediately following the next `await`able action.

I also agree, a compiler-generated state machine would likely be the only way to cleanly implement behavior like this.

@svick Agreed, where the composition is simple and doesn't branch.  But once you introduce conditional logic, error handling, looping, disposable resources, etc., the ability to neatly compose a pipeline using LINQ/Rx/callbacks becomes an absolute mess and where something like this would be useful.

**Comment by alrz** [comment]
@HaloFour IMO `foreach` on `IAsyncEnumerable<T>` shouldn't do the "immediate invoke" like that and it should be the case only for the `fork` loop. Don't you think?

**Comment by HaloFour** [comment]
@alrz Probably, that would be the benefit over a `fork` loop over `foreach`.

**Comment by TonyValenti** [comment]
This reminds me of something that I've done with PLinq in order to do async processing.
````
public TResult DoWork(TValue Value){
////Do some work with Value and return a TResult
}

....
var MyList = ...;

var MyResults = from x in MyList.AsParallel() select DoWork(x);

````

**Comment by alrz** [comment]
@TonyValenti Your `DoWork` function isn't even `async` and as it suggests, it's *parallel*, this is not about parallel loops, rather it will work side-by-side with `await` and whatever it allows you to do. Also, `AsParrallel` doesn't make sense for `IAsyncEnumerable<T>` or `IObservable<T>`, does it?

**Comment by alrz** [comment]
You can limit the degree of concurrently with the built-in `SemaphoreSlim` class,
```cs
SemaphoreSlim sem = new SemaphoreSlim(concurrentDegree);
...
fork(var item in list) {
  await sem.WaitAsync();
  // ...
  sem.Release();
}
```
If that's ever desirable.

**Comment by playsomethingsaxman** [comment]
I think you're close to missing something much more powerful: why not an `await fork` keyword pair that duplicates the task state object (entirely in heap) and continues each task after the fork, POSIX-style?

**Comment by alrz** [comment]
Moved to https://github.com/dotnet/csharplang/issues/390
